cmd/ctlog/createctconfig/main.go

```diff
@@ -111,113 +119,81 @@ func main() {
 		}
-	}
-	// Extract public component.
-	pub := key.Public()
 
-	// Encode private key to PKCS#1 ASN.1 PEM.
-	block := &pem.Block{
-		Type:  "RSA PRIVATE KEY",
-		Bytes: x509.MarshalPKCS1PrivateKey(key),
-	}
-	// Encrypt the pem
-	block, err = x509.EncryptPEMBlock(rand.Reader, block.Type, block.Bytes, []byte(*keyPassword), x509.PEMCipherAES256) // nolint
-	if err != nil {
-		logging.FromContext(ctx).Panicf("Failed to encrypt private key: %v", err)
-	}
+		if err := secret.ReconcileSecret(ctx, *secretName, ns, configMap, nsSecret); err != nil {
+			logging.FromContext(ctx).Fatalf("Failed to reconcile secret: %v", err)
+		}
 
-	privPEM, err := pem.EncodeToMemory(block), nil
-	if err != nil {
-		logging.FromContext(ctx).Panicf("Failed to encode encrypted private key: %v", err)
-	}
-	// Encode public key to PKCS#1 ASN.1 PEM.
-	pubPEM := pem.EncodeToMemory(
-		&pem.Block{
-			Type:  "RSA PUBLIC KEY",
-			Bytes: x509.MarshalPKCS1PublicKey(pub.(*rsa.PublicKey)),
-		},
-	)
-
-	// Fetch only root certificate from the chain
-	certs, err := cryptoutils.UnmarshalCertificatesFromPEM(root.ChainPEM)
-	if err != nil {
-		logging.FromContext(ctx).Panicf("unable to unmarshal certficate chain: %v", err)
-	}
-	rootCertPEM, err := cryptoutils.MarshalCertificateToPEM(certs[len(certs)-1])
-	if err != nil {
-		logging.FromContext(ctx).Panicf("unable to marshal root certificate: %v", err)
+		pubData := map[string][]byte{publicKey: configMap[publicKey]}
+		if err := secret.ReconcileSecret(ctx, *pubKeySecretName, ns, pubData, nsSecret); err != nil {
+			logging.FromContext(ctx).Panicf("Failed to reconcile public key secret %s/%s: %v", ns, *secretName, err)
+		}
+
+		logging.FromContext(ctx).Infof("Created CTLog configuration")
+		os.Exit(0)
 	}
 
-	data := map[string][]byte{
-		"private": privPEM,
-		"public":  pubPEM,
-		"rootca":  rootCertPEM,
+	// Prefer the secret config if it exists, but if it doesn't use
+	// configmap for backwards compatibility / migration.
+	if existingSecret.Data[configKey] != nil {
+		logging.FromContext(ctx).Infof("Found existing config in the secret, using that %s/%s", ns, *secretName)
+	} else {
+		existingSecret.Data[configKey] = existingCMConfig
 	}
 
-	// If the config is supposed to be written into a secret, make it so.
-	if *configInSecret {
-		data[configKey] = marshalledConfig
+	existingConfig, err := ctlog.Unmarshal(ctx, existingSecret.Data)
+	if err != nil {
+		log.Fatalf("Failed to unmarshal existing configuration: %v", err)
 	}
 
-	nsSecret := clientset.CoreV1().Secrets(ns)
-	if err := secret.ReconcileSecret(ctx, *secretName, ns, data, nsSecret); err != nil {
+	// Finally add Fulcio to it, marshal and write out.
+	existingConfig.AddFulcioRoot(ctx, root.ChainPEM)
+	marshaled, err := existingConfig.MarshalConfig(ctx)
+	if err != nil {
+		log.Fatalf("Failed to marshal new configuration: %v", err)
+	}
+	// Take out the public / private key from the secret since we didn't mess
```

Comment:
I'm unclear on why these are removed from the config?

Reply:
Ah, because the reconcilesecret only 'reconciles' the keys that are in this map. We pull them out of the map so that they are never modified by the run of this script.
https://github.com/sigstore/scaffolding/blob/main/pkg/secret/secret.go#L28